### PR TITLE
2683 name merge contact trace

### LINKED ISSUE #2683: Name merging: contact_trace
## Is your feature request related to a problem? Please describe 👀

We need to replace all FKs to `name.id` with FKs to `name_link.id` instead, to allow easy merge management.

## Describe the solution you'd like 🎁

Base: [feature/merging](https://github.com/msupply-foundation/open-msupply/tree/feature/merging)
Using the PR for [name_link + invoice](https://github.com/msupply-foundation/open-msupply/issues/2452) as a guide:
- [ ] Add `contact_trace.patient_link_id` and migrate with `contact_trace.patient_link_id = contact_trace.patient_id`
- [ ] Remove `contact_trace.patient_id`
- [ ] Update any views that relied on relations of `contact_trace.patient_id`
- [ ] Update `contact_trace` at the repository layer to return related `name` records based on `contact_trace.patient_link_id` rather than `contact_trace.patient_id`
- [ ] Add `contact_trace.contact_patient_link_id` and migrate with `contact_trace.contact_patient_link_id = contact_trace.patient_id`
- [ ] Remove `contact_trace.patient_id`
- [ ] Update any views that relied on relations of `contact_trace.patient_id`
- [ ] Update `contact_trace` at the repository layer to return related `name` records based on `contact_trace.contact_patient_link_id` rather than `contact_trace.patient_id`

### Moneyworks Jobcode 🧰

OMS:MERGE

### PR DESCRIPTION
**Pull request by clemens-msupply**
Fixes #2683 

# 👩🏻‍💻 What does this PR do? 
 
This was tricky one because the contact_trace table has a reference to two name_id and the used diesel version doesn't support joining a table twice. The solution was to create view that does all the joining. On the plus side the Rust changes are comparatively small and the link logic is completely hidden in the repository layer.

Added some mock functionality to update name_links (simulate merges). Just saw in another PR that there was something similar for items (?). Might have to align it, let me know what you think.

Also did some clean up where name_link were updated twice...

### CODE REVIEW

File: server/repository/src/tests.rs [hunk 1]
```diff
@@ -571,9 +570,6 @@ mod repository_test {
         // setup
         let name_repo = NameRowRepository::new(&connection);
         name_repo.insert_one(&data::name_1()).await.unwrap();
-        NameLinkRowRepository::new(&connection)
-            .upsert_one(&mock_name_link_from_name(&data::name_1()))
-            .unwrap();
```
Comment (clemens-msupply) [inline server/repository/src/tests.rs:576]: name_repo.insert_one already does this...

File: server/repository/src/migrations/v1_06_00/contact_trace_link_id.rs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+            ct.program_id AS program_id,
+            ct.document_id AS document_id,
+            ct.datetime AS datetime,
+            ct.contact_trace_id AS contact_trace_id,
+            patient_name_link.name_id AS patient_id,
+            contact_patient_name_link.name_id AS contact_patient_id,
+            ct.first_name AS first_name,
+            ct.last_name AS last_name,
+            ct.gender AS gender,
+            CAST(ct.date_of_birth AS DATE) AS date_of_birth,
```
Comment (clemens-msupply) [inline server/repository/src/migrations/v1_06_00/contact_trace_link_id.rs:72]: This took me quite a while to figure out. For some reason the date_of_birth column in the view changes from 4 bytes to 8 bytes  (DATE -> TIMESTAMP?) which causes Diesel to crash. By casting it back to a DATE this is fixed. Unfortunately this breaks sqlite for which reason there are two version of this view :(

File: server/repository/src/tests.rs (hunk 1 above)
Comment (Chris-Petty) [reply server/repository/src/tests.rs:576]: Nice catch!

**Review by Chris-Petty — APPROVED** [review]
Looks mint! Thanks

**Pull request merged**